ipykernel/kernelapp.py

```diff
@@ -229,9 +249,9 @@ def init_connection_file(self):
     def init_sockets(self):
         # Create a context, a session, and the kernel sockets.
         self.log.info("Starting the kernel at pid: %i", os.getpid())
-        context = zmq.Context.instance()
+        context = zmq.Context()
         # Uncomment this to try closing the context.
-        # atexit.register(context.term)
+        atexit.register(context.term)
```

Comment:
This is the source of the hang. Calling `context.term` in atexit is terminating the context prior to closing sockets, which causes a hang. This isn't safe to do without closing all sockets first. While this is a good thing to do, I don't think you need to do it in this PR. We would have to safely notify all sockets in all threads to be closed and then finally terminate the context.

```suggestion
        # atexit.register(context.term)
```

Reply:
@minrk Makes sense. If I turn that back on I get Kernel Stopped Responding in the async tests. Though it seems more just tests running later in the pool of tests start failing with timeouts. It seemed like the tests weren't cleaning up sockets and then later tests are trying to find a free socket and timing out. Running any test individually passes but the full `pytest` pass starts failing (see build for first commit of PR). Ideas on what needs to be done there?